src/nvim/api/buffer.c

```diff
@@ -1427,3 +1428,30 @@ void buf_collect_lines(buf_T *buf, size_t n, linenr_T start, int start_idx, bool
     push_linestr(lstate, l, bufstr, bufstrlen, start_idx + (int)i, replace_nl, arena);
   }
 }
+
+/// Gets current user input to prompt without invoking prompt_callback
+///
+/// @param buffer           Buffer id, or 0 for current buffer
+/// @param opts      Optional parameters
+///                  (reserved for future updates)
+/// @param[out] err         Error details, if any
+/// @return string of current prompt, or empty string for unloaded buffer / non pormpt buffer.
+String nvim_buf_get_prompt_text(Buffer buffer, Dict(option) *opts, Error *err)
+  FUNC_API_SINCE(14)
+{
+  String rv = STRING_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  // return sentinel value if the buffer isn't loaded
+  if (!buf || buf->b_ml.ml_mfp == NULL) {
+    return rv;
+  }
+
+  // check if the given buffer is prompt buffer
+  VALIDATE(bt_prompt(buf), "%s", "Not a prompt buffer", {
+    return rv;
+  });
+
+  char *current_prompt = get_current_prompt(buf);
```

Comment:
Who clears this memory? Does it get auto-cleared when the API return value goes out of scope? ASAN doesn't mind, I'm guessing it's auto-tracked somewhere.

Reply:
You need to add `FUNC_API_RET_ALLOC`